openedx/openedx-platform
# Cohort discussion settings and roles management API

This PR implements an alternative cohort discussion settings and cohort discussion roles management API for server-to-server communication. This is related to PR https://github.com/edx/edx-platform/pull/18975 PR which implements a few other API endpoints for managing cohorts. The endpoints support OAuth2 authentication and session authentication and build upon existing API endpoints for the purpose.

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:
* Log in to the LMS as a staff user.
* Create a test course and enable cohorts for it.
* Create some test users.
* Create a few cohorts and enroll the test users at random in the test course into the created cohorts.
* Enable the 'Course Discussions' feature for the course.
* Test and verify the course discussion settings endpoints using the format of the following example requests. More details are available in the view docstring for these endpoints.

        # Get the discussion settings for the given course.
        GET /api/discussion/v1/courses/{course_id}/settings

        # Example response
        {
            "always_divide_inline_discussions": false,
            "divided_inline_discussions": [],
            "divided_course_wide_discussions": [],
            "id": 1,
            "division_scheme": "cohort",
            "available_division_schemes": ["cohort"]
        }

        # Update the discussion settings for the given course.
        PATCH /api/discussion/v1/courses/{course_id}/settings
        {
            "always_divide_inline_discussions": true,
        }
        # Example response
        {
            "always_divide_inline_discussions": true,
            "divided_inline_discussions": [],
            "divided_course_wide_discussions": [],
            "id": 1,
            "division_scheme": "cohort",
            "available_division_schemes": ["cohort"]
        }

        # Get the member users for a given course and discussion role (URL encoded if necessary) - moderator, group moderator or community TA.
        GET /api/discussion/v1/courses/{course_id}/roles/{discussion rolename}

        # Example response
        {
            "results": [
                {
                    "username": "user1",
                    "email": "user1@example.com",
                    "first_name" "first name",
                    "last_name": "last name",
                    "group_name": "cohort1"
                },
                {
                    "username": "user2",
                    "email": "user2@example.com",
                    "first_name" "first name",
                    "last_name": "last name",
                    "group_name": "cohort2"
                }
            ]
        }

       # Update the discussion role membership of a user in the given course.
        POST /api/discussion/v1/courses/{course_id}/roles/{discussion rolename}

        {
            "user_id": "user1",
            "action": "revoke"
        }

        # Example response
        {
            "results": [
                {
                    "username": "user2",
                    "email": "user2@example.com",
                    "first_name" "first name",
                    "last_name": "last name",
                    "group_name": "cohort2"
                }
            ]
        }


Reviewers
